Feature/block59 Library Package pipeline for devon4node (#71)
devonfw/hangar: documentation/azure-devops/setup-library-package-pipeline.asciidoc

:idprefix:
:idseparator: -

> = Setting up a library package pipeline using Azure DevOps

Reviewer: ```suggestion
= Setting up a library package pipeline on Azure DevOps
```

> In this section we will create a package pipeline to generate a distributable of a library and push it to a feed in Azure DevOps. This pipeline will be configured in order to be triggered every time quality pipeline is executed successfully on a commit for `release/*` and `develop` branches, requiring manual launch for other branches but still enforcing that quality pipeline has passed. When Java is used, it also consumes the artifact produced by the build pipeline.

Reviewer: ```suggestion
In this section we will create a package pipeline to generate a distributable of a library and push it to a feed in Azure DevOps. This pipeline will be configured in order to be triggered every time quality pipeline is executed successfully on a commit for `release/*` and `develop` branches, requiring manual launch for other branches but still enforcing that quality pipeline has passed. Depending on the language, it consumes the artifact produced by the build pipeline.
```

```

==== Node.js project
```
./pipeline_generator.sh -c ./templates/library-package/library-package-pipeline.cfg -n library-package-pipeline -l node -d C:/Users/$USERNAME/Desktop/node-library-project --build-pipeline-name node-library-build --quality-pipeline-name node-library-quality
```

== Adding feed to library repository list

=== Maven feed
Go to Artifacts section in Azure DevOps and select the newly created Maven feed by the script (default name is `maven-feed`). Then follow the link:https://docs.microsoft.com/en-us/azure/devops/artifacts/maven/pom-and-settings?view=azure-devops[Set up the Maven client] guide from Azure Devops official documentation to perform required steps on the library project code side for being able to push the library from the pipeline.

IMPORTANT: Ignore the settings.xml part as it is created by the script. Only pom.xml modification is needed.

=== npm feed
Add a .npmrc to your project, in the same directory as your package.json:

```
registry=https://pkgs.dev.azure.com/<yourOrganization>/_packaging/<yourFeed>/npm/registry/

always-auth=true
```

> 

Author: Ok, in retrospective this seems very obvious. If you have any suggestions for naming conventions that would be helpful, as I would need 2 different ".npmrc.template" files. In the meantime I'll go with ".npmrc.project.template" for this one and ".npmrc.user.template" for the user-level npmrc.

> 

Reviewer 2: The naming convention you chose seems proper for me :)

./pipeline_generator.sh -c ./templates/library-package/library-package-pipeline.cfg -n library-package-pipeline -l java -d C:/Users/$USERNAME/Desktop/java-library-project --build-pipeline-name java-library-build --quality-pipeline-name java-library-quality
```

==== Node.js project
```

> ./pipeline_generator.sh -c ./templates/library-package/library-package-pipeline.cfg -n library-package-pipeline -l node -d C:/Users/$USERNAME/Desktop/node-library-project --build-pipeline-name node-library-build --quality-pipeline-name node-library-quality

Reviewer 3: in the pipeline-generator.sh some of the argument are not written in the help function (I think it is an error that comes from before, but better adding them)

> ./pipeline_generator.sh -c ./templates/library-package/library-package-pipeline.cfg -n library-package-pipeline -l node -d C:/Users/$USERNAME/Desktop/node-library-project --build-pipeline-name node-library-build --quality-pipeline-name node-library-quality

Reviewer 3: If I am not wrong I am not sure you need to have the build pipeline as an argument. Because we use the --build-pipeline-name in case of a java library so that we can download the application that have been built before, in the node case I am not sure this is needed. (in that case the yml.template needs to be updated with conditional step)
